functional-tests/test-bed/src/test/java/org/arquillian/smart/testing/ftest/TestBedTemplate.java

```diff
@@ -0,0 +1,91 @@
+package org.arquillian.smart.testing.ftest;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import org.arquillian.smart.testing.ftest.testbed.project.Project;
+import org.arquillian.spacelift.Spacelift;
+import org.arquillian.spacelift.process.CommandBuilder;
+import org.arquillian.spacelift.task.os.CommandTool;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestName;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * This template is responsible for cloning source repository on which smart testing
+ * provider will be tested and providing access to underlying project for further manipulations
+ * and build execution.
+ *
+ * For each test method original repository which is cloned from defined location will be copied
+ * to a new folder named after following pattern: [original-repo-name]_[class name]_[method_name].
+ *
+ * Each test has access to its own {@link Project} instance which allows to apply desired changes
+ * before running the build.
+ */
+public abstract class TestBedTemplate {
+
+    private static final String ORIGIN = "https://github.com/arquillian/smart-testing-dogfood-repo.git";
+    private static final String REPO_NAME = ORIGIN.substring(ORIGIN.lastIndexOf('/') + 1).replace(".git", "");
+
+    @ClassRule
+    public static final TemporaryFolder TMP_FOLDER = new TemporaryFolder();
+
+    @Rule
+    public final TestName name = new TestName();
+
+    private static String GIT_REPO_FOLDER;
+
+    protected Project project;
```

Comment:
Yeah, I was thinking about the rule too, it would serve the same purpose but different way of composing. But I'm not sure if we really have to do it as part of this PR. Might be something to consider as part of #56 ?